src/graph/GoExecutor.cpp

```diff
@@ -723,85 +708,207 @@ void GoExecutor::finishExecution(RpcResponse &&rpcResp) {
+        }
+        if (!ret.ok()) {
+            LOG(ERROR) << "Get rows failed: " << ret.status();
+            return;
+        }
+        if (!ret.value().empty()) {
             resp_->set_rows(std::move(ret).value());
         }
     }
     doFinish(Executor::ProcessControl::kNext);
 }
 
-StatusOr<std::vector<storage::cpp2::PropDef>> GoExecutor::getStepOutProps() {
-    std::vector<storage::cpp2::PropDef> props;
-    for (auto &e : edgeTypes_) {
-        storage::cpp2::PropDef pd;
-        pd.owner = storage::cpp2::PropOwner::EDGE;
-        pd.name = _DST;
-        pd.id.set_edge_type(e);
-        props.emplace_back(std::move(pd));
-        // We need ranking when go reverly in final step,
-        // because we have to fetch the coresponding edges.
-        if (isReversely() && isFinalStep()) {
-            storage::cpp2::PropDef rankPd;
-            rankPd.owner = storage::cpp2::PropOwner::EDGE;
-            rankPd.name = _RANK;
-            rankPd.id.set_edge_type(e);
-            props.emplace_back(std::move(rankPd));
+
+bool GoExecutor::onlyStructure() {
+    return whereWrapper_->filter_  == nullptr
+            && edgeTypes_.size() == 1
+            && (yields_.empty()
+                || (yields_.size() == 1 && yields_[0]->expr()->isEdgeDstIdExpression()));
+}
+
+StatusOr<std::vector<cpp2::RowValue>> GoExecutor::toThriftResponse(RpcResponse&& rpcResp) {
+    std::vector<cpp2::RowValue> rows;
+    int64_t totalRows = 0;
+    for (auto& resp : rpcResp.responses()) {
+        if (resp.get_total_edges() != nullptr) {
+            totalRows += *resp.get_total_edges();
         }
     }
-
-    if (!isFinalStep()) {
-        return props;
+    rows.reserve(totalRows);
+    if (onlyStructure()) {
+        auto uniqResult = std::make_unique<std::unordered_set<VertexID>>();
+        auto& all = rpcResp.responses();
+        for (auto &resp : all) {
+            if (resp.get_vertices() == nullptr) {
+                continue;
+            }
+            for (auto& vdata : resp.vertices) {
+                DCHECK(vdata.__isset.edge_data);
+                for (auto& edata : vdata.edge_data) {
+                    for (auto& edge : edata.edges) {
+                        // Check if duplicate
+                        if (distinct_) {
+                            auto ret = uniqResult->emplace(edge.get_dst());
+                            if (!ret.second) {
+                                continue;
+                            }
+                        }
+                        std::vector<cpp2::ColumnValue> row;
+                        row.emplace_back();
+                        row.back().set_id(edge.get_dst());
+                        rows.emplace_back();
+                        rows.back().set_columns(std::move(row));
+                    }  // for edges
+                }  // for edata
+            }   // for `vdata'
+        }   // for `resp'
+        return rows;
```

Comment:
maybe define this function in a sub function is more readable, else the function is too long